streaming/src/test/scala/org/apache/spark/streaming/ReceivedBlockHandlerSuite.scala

```diff
@@ -174,6 +186,120 @@ class ReceivedBlockHandlerSuite
+    // IteratorBlock-MEMORY_AND_DISK
+    testRecordcount(false, StorageLevel.MEMORY_AND_DISK,
+      IteratorBlock((ArrayBuffer.fill(150)(0)).iterator), blockManager, Some(150))
+  }
+
+  test("BlockManagerBasedBlockHandler-MEMORY_ONLY - isFullyConsumed") {
+    storageLevel = StorageLevel.MEMORY_ONLY
+    blockManager = createBlockManager(12000)
+    val block = List.fill(70)(new Array[Byte](100))
+    // spark.storage.unrollFraction set to 0.4 for BlockManager
+    // With 12000 * 0.4 = 4800 bytes of free space for unroll, there is not enough space to store
+    // this block With MEMORY_ONLY StorageLevel. BlockManager will not be able to unroll this block
+    // and hence it will not tryToPut this block, resulting the SparkException
+    withBlockManagerBasedBlockHandler { handler =>
+      val thrown = intercept[SparkException] {
+        val blockStoreResult = storeBlock(handler, IteratorBlock(block.iterator))
+      }
+      assert(thrown.getMessage ===
+        "Could not store input-1-1000 to block manager with storage level " + storageLevel)
+    }
+  }
+
+  test("BlockManagerBasedBlockHandler-MEMORY_AND_DISK - isFullyConsumed") {
+    blockManager = createBlockManager(12000)
+    // spark.storage.unrollFraction set to 0.4 for BlockManager
+    // With 12000 * 0.4 = 4800 bytes of free space for unroll, there is not enough space to store
+    // this block in MEMORY, But BlockManager will be able to sereliaze this block to DISK
+    // and hence count returns correct value.
+    testRecordcount(true, StorageLevel.MEMORY_AND_DISK,
+      IteratorBlock((List.fill(70)(new Array[Byte](100))).iterator), blockManager, Some(70))
+  }
+
+  test("WriteAheadLogBasedBlockHandler-MEMORY_ONLY - isFullyConsumed") {
+    blockManager = createBlockManager(12000)
+    // spark.storage.unrollFraction set to 0.4 for BlockManager
+    // With 12000 * 0.4 = 4800 bytes of free space for unroll, there is not enough space to store
+    // this block in MEMORY, But BlockManager will be able to sereliaze this block to WAL
+    // and hence count returns correct value.
+    testRecordcount(false, StorageLevel.MEMORY_ONLY,
+      IteratorBlock((List.fill(70)(new Array[Byte](100))).iterator), blockManager, Some(70))
+  }
+
+  /**
+   * Test storing of data using different types of Handler, StorageLevle and ReceivedBlocks
+   * and verify the correct record count
+   */
+  private def testRecordcount(isBlockManagedBasedBlockHandler: Boolean,
+      sLevel: StorageLevel,
+      receivedBlock: ReceivedBlock,
+      bManager: BlockManager,
+      expectedNumRecords: Option[Long]
+      ) {
+   storageLevel = sLevel
+   blockManager = bManager
+   if (isBlockManagedBasedBlockHandler) {
+      // test received block with BlockManager based handler
+      withBlockManagerBasedBlockHandler { handler =>
+        val blockStoreResult = storeBlock(handler, receivedBlock)
+        assert(blockStoreResult.numRecords === expectedNumRecords)
```

Comment:
done..message will come like this in case of failure..

Some(25) did not equal Some(21) Message count not matches for a org.apache.spark.streaming.receiver.ArrayBufferBlock block being inserted using BlockManagerBasedBlockHandler with StorageLevel(false, true, false, true, 1) (ReceivedBlockHandlerSuite.scala:274)